pkg/simple/client/factory.go

```diff
@@ -0,0 +1,143 @@
+	redisOptions      *redis.RedisOptions
+	kubernetesOptions *k8s.KubernetesOptions
+	devopsOptions     *devops.DevopsOptions
+	ldapOptions       *ldap.LdapOptions
+	s3Options         *s2is3.S3Options
+}
+
+func NewClientSetOptions() *ClientSetOptions {
+	return &ClientSetOptions{
+		mySQLOptions:      mysql.NewMySQLOptions(),
+		redisOptions:      redis.NewRedisOptions(),
+		kubernetesOptions: k8s.NewKubernetesOptions(),
+		ldapOptions:       ldap.NewLdapOptions(),
+		devopsOptions:     devops.NewDevopsOptions(),
+		s3Options:         s2is3.NewS3Options(),
+	}
+}
+
+func (c *ClientSetOptions) SetMySQLOptions(options *mysql.MySQLOptions) *ClientSetOptions {
+	c.mySQLOptions = options
+	return c
+}
+
+func (c *ClientSetOptions) SetRedisOptions(options *redis.RedisOptions) *ClientSetOptions {
+	c.redisOptions = options
+	return c
+}
+
+func (c *ClientSetOptions) SetKubernetesOptions(options *k8s.KubernetesOptions) *ClientSetOptions {
+	c.kubernetesOptions = options
+	return c
+}
+
+func (c *ClientSetOptions) SetDevopsOptions(options *devops.DevopsOptions) *ClientSetOptions {
+	c.devopsOptions = options
+	return c
+}
+
+func (c *ClientSetOptions) SetLdapOptions(options *ldap.LdapOptions) *ClientSetOptions {
+	c.ldapOptions = options
+	return c
+}
+
+func (c *ClientSetOptions) SetS3Options(options *s2is3.S3Options) *ClientSetOptions {
+	c.s3Options = options
+	return c
+}
+
+// ClientSet provide best of effort service to initialize clients,
+// but there is no guarantee to return a valid client instance,
+// so do validity check before use
+type ClientSet struct {
+	mySQLClient  *mysql.MySQLClient
+	k8sClient    *k8s.KubernetesClient
+	ldapClient   *ldap.LdapClient
+	devopsClient *devops.DevopsClient
+	redisClient  *redis.RedisClient
+	s3Client     *s2is3.S3Client
+}
+
+// global clientsets instance
+var sharedClientSet *ClientSet
+
+func ClientSets() *ClientSet {
+	return sharedClientSet
+}
+
+func NewClientSetFactory(c *ClientSetOptions, stopCh <-chan struct{}) *ClientSet {
+	sharedClientSet = &ClientSet{}
+
+	// create clients one by one
+	if c.mySQLOptions != nil && c.mySQLOptions.Host != "" {
+		sharedClientSet.mySQLClient = mysql.NewMySQLClient(c.mySQLOptions, stopCh)
+	}
+
+	if c.kubernetesOptions != nil {
+		sharedClientSet.k8sClient = k8s.NewKubernetesClient(c.kubernetesOptions)
+	}
+
+	if c.ldapOptions != nil && c.ldapOptions.Host != "" {
+		sharedClientSet.ldapClient = ldap.NewLdapClient(c.ldapOptions, stopCh)
+	}
+
+	if c.devopsOptions != nil && c.devopsOptions.Host != "" {
+		sharedClientSet.devopsClient = devops.NewDevopsClient(c.devopsOptions)
+	}
+
+	if c.redisOptions != nil && c.redisOptions.Host != "" {
+		sharedClientSet.redisClient = redis.NewRedisClient(c.redisOptions, stopCh)
+	}
+
+	if c.s3Options != nil && c.s3Options.Endpoint != "" {
+		sharedClientSet.s3Client = s2is3.NewS3Client(c.s3Options)
+	}
+
+	return sharedClientSet
+}
+
+// no guarantee to return a valid connection
+func (cs *ClientSet) MySQL() *mysql.Database {
```

Comment:
The idea now is to use the Client is nil to determine whether the function is enabled?
In the wrong configuration, Client will also be nil.